### PR TITLE
Mobile 3936 - Inline reminders

### PR DESCRIPTION
**Pull request by crazyserver**

### CODE REVIEW

**Review by dpalou — CHANGES_REQUESTED** [review]
Thanks Pau, I left a lot of comments here but they're all minor things, the solution looks good to me.

About the tests, I missed some things to be tested:

- We aren't testing that the "Reminder set" toast displays the right date. You could create an activity/course with a fixed time (e.g. 1/1/2050) and that way you can know for sure the date that will be displayed in the Toast.
- We aren't testing that local notifications can be canceled (they aren't triggered). 

These things only need to be tested once, there's no need to replicate them for course and activity.

File: src/core/services/sites.ts [hunk 1]
```diff
@@ -1633,26 +1647,37 @@ export class CoreSitesProvider {
...
         const promise = site.getDb().createTableFromSchema(SCHEMA_VERSIONS_TABLE_SCHEMA)
             .then(() => this.applySiteSchemas(site, this.siteSchemas));
 
-        this.siteSchemasMigration[site.id] = promise;
+        this.siteSchemasMigration[siteId] = promise;
 
-        return promise.finally(() => {
+        return promise.then(() => {
+            // Set site after migrating schemas, or a call to getSite could get the site while tables are being created.
+            this.sites[siteId] = site;
```
[the remark this answers is not in the document]
Comment (crazyserver) [reply src/core/services/sites.ts:1677]: You're right!

File: src/addons/calendar/services/calendar.ts [hunk 2]
```diff
@@ -1429,95 +1437,25 @@ export class AddonCalendarProvider {
      *
      * @return Promise resolved when all the notifications have been scheduled.
      */
-    async scheduleAllSitesEventsNotifications(): Promise<void> {
+    async updateAllSitesEventReminders(): Promise<void> {
```
[the remark this answers is not in the document]
Comment (crazyserver) [reply src/addons/calendar/services/calendar.ts:1365]: Done thanks!

File: src/core/features/reminders/services/database/reminders.ts [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+    await db.insertRecord(REMINDERS_TABLE, reminder);
+};
+
+export type CoreReminderDBRecord = {
+    id: number; // Reminder ID.
+    component: string; // Component where the reminder belongs.
+    instanceId: number; // Instance Id where the reminder belongs.
+    type: string; // Event idenfier type.
+    time: number; // Event time.
+    before: number; // Seconds before the event to remind.
```
[the remark this answers is not in the document]
Comment (crazyserver) [reply src/core/features/reminders/services/database/reminders.ts:236]: Ok thanks!

File: src/core/features/reminders/services/reminders.ts [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+        const siteIds = await CoreSites.getSitesIds();
+
+        await Promise.all(siteIds.map((siteId: string) => async () => {
+            const reminders = await this.getAllReminders(siteId);
+
+            reminders.forEach((reminder) => {
+                this.scheduleNotification(reminder, siteId);
+            });
+        }));
+    }
```
[the remark this answers is not in the document]
Comment (crazyserver) [reply src/core/features/reminders/services/reminders.ts:334]: So, when the app starts, the events have been always scheculed before?

[the remark this answers is not in the document]
Comment (dpalou) [reply src/core/features/reminders/services/reminders.ts:334]: Well, if a reminder already exists in DB it means that the notification was scheduled before, right? And I think there's no need to reschedule them again, Android should keep the scheduled notifications.

You can do a test if you want: remove this code, create a couple of reminders to be triggered soon, kill the app and restart it and check if the notifications are triggered.

File: src/addons/calendar/services/calendar.ts [hunk 5]
```diff
@@ -103,8 +102,7 @@ export class AddonCalendarProvider {
 
     static readonly DAYS_INTERVAL = 30;
     static readonly COMPONENT = 'AddonCalendarEvents';
-    static readonly DEFAULT_NOTIFICATION_TIME_CHANGED = 'AddonCalendarDefaultNotificationTimeChangedEvent';
-    static readonly DEFAULT_NOTIFICATION_TIME_SETTING = 'mmaCalendarDefaultNotifTime';
```
[the remark this answers is not in the document]
Comment (crazyserver) [reply src/addons/calendar/services/calendar.ts:112]: Got it! Thanks!

File: src/core/features/reminders/components/set-reminder-menu/set-reminder-menu.ts [hunk 6]
```diff
@@ -149,6 +174,14 @@ export class CoreRemindersSetReminderMenuComponent implements OnInit {
         // Let the dimissed popover to be removed.
         await CoreUtils.nextTick();
 
+        if (this.eventTime) {
+            if (!this.isValidTime(this.customUnits, Math.abs(this.customValue))) {
+                PopoverController.dismiss();
+
+                return;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (crazyserver) [reply src/core/features/reminders/components/set-reminder-menu/set-reminder-menu.ts:183]: Thanks I've totally removed the conditions so it saves the reminder

File: src/core/features/emulator/services/local-notifications.ts [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+        if (!nextNotification) {
+            return;
+        }
+
+        const notificationTime = nextNotification.trigger?.at?.getTime() || 0;
+        const timeout = notificationTime - Date.now();
+        if (timeout <= 0) {
+            this.triggerNextNotification();
+
+            return;
```
[the remark this answers is not in the document]
Comment (crazyserver) [reply src/core/features/emulator/services/local-notifications.ts:86]: It is a simplification,t this function checks the next notification and triggers it when notification time has been reached.

File: src/core/features/reminders/services/reminders.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (crazyserver) [reply src/core/features/reminders/services/reminders.ts:334]: Ok, I've added a new commit that we should discuss

[the remark this answers is not in the document]
Comment (crazyserver) [reply src/core/features/reminders/services/reminders.ts]: Discussed and merged on the LocalNotifications mock commit

File: src/core/features/reminders/tests/behat/activity_reminders.feature [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    Given the following "users" exist:
+      | username | firstname | lastname | email                |
+      | student1 | Student   | student  | student1@example.com |
+    And the following "courses" exist:
+      | fullname | shortname | category |
+      | Course 1 | C1        | 0        |
+    And the following "course enrolments" exist:
+      | user     | course | role           |
+      | student1 | C1     | student        |
+    Given the following "activities" exist:
```
Comment (dpalou) [inline src/core/features/reminders/tests/behat/activity_reminders.feature:14]:
```suggestion
    And the following "activities" exist:
```

File: src/core/features/reminders/tests/behat/activity_reminders.feature [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    And I press "Custom..." in the app
+    Then I should find "Custom reminder" in the app
+    When I set the following fields to these values in the app:
+       | Value | 68 |
+       | Units | minutes |
+    And I press "Set reminder" in the app
+    Then I should find "Reminder set for " in the app
+    When I press "Set a reminder for \"Assignment 01\" (Due)" in the app
+    Then I should find "Reminder set for " in the app
+    And I press "Delete reminder" in the app
```
Comment (dpalou) [inline src/core/features/reminders/tests/behat/activity_reminders.feature:88]:
```suggestion
    When I press "Delete reminder" in the app
```

File: src/core/features/reminders/tests/behat/activity_reminders.feature [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+    When I set the following fields to these values in the app:
+       | Value | 69 |
+       | Units | minutes |
+    And I press "Set reminder" in the app
+    Then I should find "Reminder set for " in the app
+    When I wait "50" seconds
+    Then a notification with title "Due: Assignment 01" is present in the app
+    And I close a notification with title "Due: Assignment 01" in the app
+
+
```
Comment (dpalou) [inline src/core/features/reminders/tests/behat/activity_reminders.feature:76]: You have an extra blank line here.

**Review by dpalou — CHANGES_REQUESTED** [review]

**Review by dpalou — APPROVED** [review]

**Comment by dpalou** [comment]
Thanks Pau, it looks good now but you have some linter errors:

https://github.com/moodlehq/moodleapp/actions/runs/3436891988/jobs/5730955597

**Pull request merged**